diff -Nur a/sommelier.cc b/sommelier.cc
--- a/sommelier.cc	2022-11-06 19:29:16.580361574 +0800
+++ b/sommelier.cc	2022-11-06 19:37:28.830367176 +0800
@@ -2546,16 +2546,13 @@
   wl_array_init(&ctx->selection_data);
   ctx->selection_data_ack_pending = 0;
 
-  int pipe_fd;
-  rv = ctx->channel->create_pipe(pipe_fd);
-  if (rv) {
-    fprintf(stderr, "error: failed to create virtwl pipe: %s\n", strerror(-rv));
-    sl_send_selection_notify(ctx, XCB_ATOM_NONE);
-    return;
-  }
+  int p[2];
+
+  rv = pipe2(p, O_CLOEXEC | O_NONBLOCK);
+  errno_assert(!rv);
 
-  fd_to_receive = pipe_fd;
-  fd_to_wayland = pipe_fd;
+  fd_to_receive = p[0];
+  fd_to_wayland = p[1];
 
   xcb_get_atom_name_reply_t* atom_name_reply =
       xcb_get_atom_name_reply(ctx->connection, atom_name_cookie, NULL);
@@ -3634,6 +3631,7 @@
   struct sl_context ctx;
   sl_context_init_default(&ctx);
 
+  const char* drm_device = getenv("SOMMELIER_DRM_DEVICE");
   const char* display = getenv("SOMMELIER_DISPLAY");
   const char* scale = getenv("SOMMELIER_SCALE");
   const char* dpi = getenv("SOMMELIER_DPI");
@@ -3746,6 +3744,8 @@
       ctx.enable_xshape = true;
     } else if (strstr(arg, "--virtgpu-channel") == arg) {
       ctx.use_virtgpu_channel = true;
+    } else if (strstr(arg, "--drm-device") == arg) {
+      drm_device = sl_arg_value(arg);
 #ifdef PERFETTO_TRACING
     } else if (strstr(arg, "--trace-filename") == arg) {
       ctx.trace_filename = sl_arg_value(arg);
@@ -3856,8 +3856,7 @@
     return EXIT_FAILURE;
   }
 
-  char* drm_device = NULL;
-  int drm_fd = open_virtgpu(&drm_device);
+  int drm_fd = open(drm_device, O_RDWR | O_CLOEXEC);
   if (drm_fd >= 0) {
     ctx.gbm = gbm_create_device(drm_fd);
     if (!ctx.gbm) {