diff -Nur a/sommelier.cc b/sommelier.cc
--- a/sommelier.cc	2022-11-06 19:29:16.580361574 +0800
+++ b/sommelier.cc	2022-11-06 19:37:28.830367176 +0800
@@ -616,7 +616,7 @@
       data_device_manager->host_global =
           sl_data_device_manager_global_create(ctx);
     }
-  } else if (strcmp(interface, "xdg_wm_base") == 0) {
+  } else if (strcmp(interface, "xdg_wm_base") == 0 || strcmp(interface, "zxdg_shell_v6") == 0) {
     struct sl_xdg_shell* xdg_shell =
         static_cast<sl_xdg_shell*>(malloc(sizeof(struct sl_xdg_shell)));
     assert(xdg_shell);
